### ISSUE TITLE
N818 ErrorSuffixOnExceptionName should apply recursively to sub-subclasses of Exception

### ISSUE THREAD

**Issue by maltevesper** [astral-sh/ruff#20936]
This is created to reraise #783, as requested by @ntBre in [#783](https://github.com/astral-sh/ruff/issues/783#issuecomment-3410784877)_

From the original issue:

> This triggers `N818 exception name 'Bar' should be named with an Error suffix` in pep8-naming but not in Ruff:
> 
> class FooError(Exception):
>     pass
> 
> class Bar(FooError):
>     pass
> pep8-naming checks this by [recursively searching](https://github.com/PyCQA/pep8-naming/blob/b3492eeede12a26d35c0ee9e6b9c37a7e5484bf6/src/pep8ext_naming.py#L283-L293) all ancestor classes, although it may be nearly as effective to just check whether an immediate parent class has a name ending in `Error`.

I think there are a few reasons to reconsider recursivly checking baseclass for `Exception` to make the lint more viable.

```
pydantic.ValidationErrr(ValueError): # typo missing O
```

or for projects defining their own Exception base class on which they base all other exceptions, one example would be [gitlab-python](https://github.com/python-gitlab/python-gitlab/blob/3859aa0ff3a77b31651b770585f5cfe81c9d244a/gitlab/exceptions.py#L37).

```
class GitlabAuthenticationError(GitlabError):
    pass


class RedirectError(GitlabError):
    pass
```

If it is still not viable it would be nice to have a rational like "prohibitve performance impact" statet for reference.
If you can point me to another lint already recursively inspecting bases I *might* have a go at a PR

**Comment by ntBre** [comment]
Thanks for spinning off the separate issue! I think this makes sense but will require type inference, or at least multi-file analysis, in the general case where base classes might come from different modules. As noted above, we currently just check if any of the immediate base classes is `Exception` or ends with `Error`:

https://github.com/astral-sh/ruff/blob/c7e2bfd7592d765df07b87ad64755d5fa9037fa4/crates/ruff_linter/src/rules/pep8_naming/rules/error_suffix_on_exception_name.rs#L65

We could try using [`any_base_class`](https://github.com/astral-sh/ruff/blob/c7e2bfd7592d765df07b87ad64755d5fa9037fa4/crates/ruff_python_semantic/src/analyze/class.rs#L29), which I think will do the recursive traversal within a single file, in the meantime though.

`N818` is a stable rule now, so if anyone gives this a try, we'll need to make it a [`preview`](https://github.com/astral-sh/ruff/blob/c0fb235a703684bca1cc1767015bb7e19ab1c22f/crates/ruff_linter/src/preview.rs#L16-L19) change.

**Comment by MichaReiser** [comment]
I think it's fine making this change as it leads to fewer diagnostics (unless it results in a very large diff of errors)

**Comment by ntBre** [comment]
I think this could actually lead to more diagnostics since we'll inspect more super classes. We only flag `FooErrr` in this [example](https://play.ruff.rs/dcef67b1-93a6-42eb-9aac-ec8788eda95d) currently, but would also flag `Bar` if we resolved its base classes recursively:

```py
class FooErrr(Exception): ...

class Bar(FooErrr): ...
```

if I'm understanding correctly.

**Comment by bitloi** [comment]
@MichaReiser Can I work on this issue?

**Comment by MichaReiser** [comment]
@bitloi sure. Can I ask you to give us some time to review and merge your open PRs? It otherwise can feel a bit overwhelming :)

**Comment by bitloi** [comment]
Sure, no problem. Thanks for your support.
